contracts/libraries/SwapToRatio.sol

```diff
@@ -0,0 +1,189 @@
+                return (false, 0, 0);
+            }
+        } else {
+            if (sqrtRatioX96Target > positionParams.sqrtRatioX96Upper) {
+                return (false, 0, 0);
+            }
+        }
+
+        uint256 token0Delta =
+            SqrtPriceMath.getAmount0Delta(
+                poolParams.sqrtRatioX96,
+                sqrtRatioX96Target,
+                poolParams.liquidity,
+                zeroForOne ? false : true
+            );
+        uint256 token1Delta =
+            SqrtPriceMath.getAmount1Delta(
+                poolParams.sqrtRatioX96,
+                sqrtRatioX96Target,
+                poolParams.liquidity,
+                zeroForOne ? true : false
+            );
+
+        uint256 validDeposit0 =
+            SqrtPriceMath.getAmount0Delta(
+                sqrtRatioX96Target,
+                positionParams.sqrtRatioX96Upper,
+                poolParams.liquidity,
+                false
+            );
+        uint256 validDeposit1 =
+            SqrtPriceMath.getAmount1Delta(
+                sqrtRatioX96Target,
+                positionParams.sqrtRatioX96Lower,
+                poolParams.liquidity,
+                false
+            );
+
+        if (zeroForOne) {
+            amount0Updated = positionParams.amount0Initial - ((token0Delta * 1e6) / (1e6 - poolParams.fee));
+            amount1Updated = positionParams.amount1Initial + token1Delta;
+            doSwap = (amount0Updated) / amount1Updated > (validDeposit0) / validDeposit1;
+        } else {
+            amount0Updated = positionParams.amount0Initial + token0Delta;
+            amount1Updated = positionParams.amount1Initial - ((token1Delta * 1e6) / (1e6 - poolParams.fee));
+            doSwap = (amount1Updated) / amount0Updated > (validDeposit1) / validDeposit0;
+        }
+    }
+
+    // TODO: address range order scenarios
+    function getPostSwapPrice(IUniswapV3Pool pool, PositionParams memory positionParams)
+        internal
+        view
+        returns (uint160)
+    {
+        (PoolParams memory poolParams, int24 tickSpacing, int24 tick) = getPoolInputs(pool);
+        bool zeroForOne =
+            isZeroForOne(
+                positionParams.amount0Initial,
+                positionParams.amount1Initial,
+                poolParams.sqrtRatioX96,
+                positionParams.sqrtRatioX96Lower,
+                positionParams.sqrtRatioX96Upper
+            );
+        bool crossTickBoundary = true;
+        uint256 amount0Next;
+        uint256 amount1Next;
+        int24 nextInitializedTick;
+        uint160 sqrtRatioNextX96;
+
+        while (crossTickBoundary) {
+            // returns the next initialized tick or the last tick within one word of the current tick.
+            // We'll renew calculation at least on a per word basis for better rounding
+            (nextInitializedTick, ) = pool.nextInitializedTickWithinOneWord(tick, tickSpacing, zeroForOne);
```

Comment:
ok, this might be really annoying, but in theory every tick of a word could be initialized, and iiuc we would be making repeat calls every time to fetch the whole word in that case? need an idempotent fetch with caching 😬😬😬